File: tests/CostFunctionTests.cpp
#include <gtest/gtest.h>

#include "clf/CostFunction.hpp"

#include "TestPenaltyFunctions.hpp"

using namespace clf;

TEST(CostFunctionTests, DenseMatrices) {
  // create the example penalty functions
  auto func0 = std::make_shared<tests::DensePenaltyFunctionTest0>();
  auto func1 = std::make_shared<tests::DensePenaltyFunctionTest1>();

  DenseCostFunction cost({func0, func1});
  EXPECT_EQ(cost.InputDimension(), func0->indim);
  EXPECT_EQ(cost.InputDimension(), func1->indim);
  EXPECT_EQ(cost.numPenaltyFunctions, 2);
  EXPECT_EQ(cost.numTerms, func0->outdim+func1->outdim);

  const Eigen::VectorXd beta = Eigen::VectorXd::Random(cost.InputDimension());
  const Eigen::VectorXd eval = cost.Evaluate(beta);
  EXPECT_EQ(eval.size(), cost.numTerms);

  Eigen::VectorXd expected(8);
  expected << beta(0), beta(0)*(1.0-beta(2)), 1.0-beta(1), 1.0-beta(1)+beta(2), beta(2), beta(2)*(1.0-beta(1)), beta(0)*beta(2), beta(0)*beta(0)*beta(1);
  EXPECT_NEAR((eval-expected).norm(), 0.0, 1.0e-14);

  const Eigen::MatrixXd jac = cost.Jacobian(beta);
  EXPECT_EQ(jac.cols(), func0->indim);
  EXPECT_EQ(jac.cols(), func1->indim);
  EXPECT_EQ(jac.rows(), func0->outdim+func1->outdim);
  EXPECT_NEAR((jac.topLeftCorner(func0->outdim, func0->indim)-func0->Jacobian(beta)).norm(), 0.0, 1.0e-14);
  EXPECT_NEAR((jac.bottomLeftCorner(func1->outdim, func1->indim)-func1->Jacobian(beta)).norm(), 0.0, 1.0e-14);

  const Eigen::VectorXd grad = cost.Gradient(eval, jac);
  EXPECT_EQ(grad.size(), func0->indim);  
  EXPECT_EQ(grad.size(), func1->indim);  
  EXPECT_NEAR((grad-cost.Gradient(beta)).norm(), 0.0, 1.0e-14);
  Eigen::VectorXd gradFD = Eigen::VectorXd::Zero(func0->indim);
  { // estimate the gradient with 8th order finite differences
    Eigen::VectorXd b = beta;
    Eigen::Vector4d weights(0.8, -0.2, 4.0/105.0, -1.0/280.0);  
    const double delta = 1.0e-2;
    for( std::size_t component=0; component<func0->indim; ++component ) {
      for( std::size_t j=0; j<weights.size(); ++j ) {
	b(component) += delta;
	const Eigen::VectorXd c = cost.Evaluate(b);
	gradFD(component) += weights(j)*c.dot(c);
      }
      b(component) -= weights.size()*delta;
      for( std::size_t j=0; j<weights.size(); ++j ) {
	b(component) -= delta;
	const Eigen::VectorXd c = cost.Evaluate(b);
	gradFD(component) -= weights(j)*c.dot(c);
      }
      b(component) += weights.size()*delta;
    }
    gradFD /= delta;
  }
  EXPECT_NEAR((grad-gradFD).norm(), 0.0, 1.0e-12);

  const Eigen::MatrixXd hess = cost.Hessian(beta);
  Eigen::MatrixXd hessFD = Eigen::MatrixXd::Zero(func0->indim, func0->indim);
  EXPECT_EQ(hess.rows(), hessFD.rows());
  EXPECT_EQ(hess.cols(), hessFD.cols());
  { // estimate the Hessian with 8th order finite differences
    Eigen::VectorXd b = beta;
    Eigen::Vector4d weights(0.8, -0.2, 4.0/105.0, -1.0/280.0);  
    const double delta = 1.0e-2;
    for( std::size_t component=0; component<func0->indim; ++component ) {
      for( std::size_t j=0; j<weights.size(); ++j ) {
	b(component) += delta;
	hessFD.col(component) += 2.0*weights(j)*cost.Jacobian(b).adjoint()*cost.Evaluate(b);
      }
      b(component) -= weights.size()*delta;
      for( std::size_t j=0; j<weights.size(); ++j ) {
	b(component) -= delta;
	hessFD.col(component) -= 2.0*weights(j)*cost.Jacobian(b).adjoint()*cost.Evaluate(b);
      }
      b(component) += weights.size()*delta;
    }
    hessFD /= delta;
  }
  EXPECT_NEAR((hess-hessFD).norm(), 0.0, 1.0e-12);
}

TEST(CostFunctionTests, SparseMatrices) {
  // create the example penalty functions
  auto func0 = std::make_shared<tests::SparsePenaltyFunctionTest0>();
  auto func1 = std::make_shared<tests::SparsePenaltyFunctionTest1>();

  SparseCostFunction cost({func0, func1});
  EXPECT_EQ(cost.InputDimension(), func0->indim);
  EXPECT_EQ(cost.InputDimension(), func1->indim);
  EXPECT_EQ(cost.numPenaltyFunctions, 2);
  EXPECT_EQ(cost.numTerms, func0->outdim+func1->outdim);

  const Eigen::VectorXd beta = Eigen::VectorXd::Random(cost.InputDimension());
  const Eigen::VectorXd eval = cost.Evaluate(beta);
  EXPECT_EQ(eval.size(), cost.numTerms);

  Eigen::VectorXd expected(8);
  expected << beta(0), beta(0)*(1.0-beta(2)), 1.0-beta(1), 1.0-beta(1)+beta(2), beta(2), beta(2)*(1.0-beta(1)), beta(0)*beta(2), beta(0)*beta(0)*beta(1);
  EXPECT_NEAR((eval-expected).norm(), 0.0, 1.0e-14);

  const Eigen::SparseMatrix<double> jac = cost.Jacobian(beta);
  EXPECT_EQ(jac.cols(), func0->indim);
  EXPECT_EQ(jac.cols(), func1->indim);
  EXPECT_EQ(jac.rows(), func0->outdim+func1->outdim);
  EXPECT_NEAR((jac.topLeftCorner(func0->outdim, func0->indim)-func0->Jacobian(beta)).norm(), 0.0, 1.0e-14);
  EXPECT_NEAR((jac.bottomLeftCorner(func1->outdim, func1->indim)-func1->Jacobian(beta)).norm(), 0.0, 1.0e-14);
}

File: tests/CostFunctionTests.py
"""! @brief Test the cost function (clf::CostFunction) python interface"""

import unittest

import numpy as np
import scipy

import random

from TestPenaltyFunctions import *
import PyCoupledLocalFunctions as clf

class TestCostFunction(unittest.TestCase):
    """! Test the cost function (clf::CostFunction) python interface"""

    def tearDown(self):
        """! Check the evaluation and derivative sizes"""
        
        self.assertEqual(self.cost.InputDimension(), self.func0.indim)
        self.assertEqual(self.cost.InputDimension(), self.func1.indim)
        self.assertEqual(self.cost.numPenaltyFunctions, 2)
        self.assertEqual(self.cost.numTerms, self.func0.outdim+self.func1.outdim)

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(self.cost.InputDimension())])

        fx = self.cost.Evaluate(beta)
        self.assertAlmostEqual(fx[0], beta[0])
        self.assertAlmostEqual(fx[1], beta[0]*(1.0-beta[2]))
        self.assertAlmostEqual(fx[2], 1.0-beta[1])
        self.assertAlmostEqual(fx[3], 1.0-beta[1]+beta[2])
        self.assertAlmostEqual(fx[4], beta[2])
        self.assertAlmostEqual(fx[5], beta[2]*(1.0-beta[1]))
        self.assertAlmostEqual(fx[6], beta[0]*beta[2])
        self.assertAlmostEqual(fx[7], beta[0]*beta[0]*beta[1])

        grad = self.cost.Gradient(beta);
        self.assertEqual(len(grad), self.func0.indim)
        self.assertEqual(len(grad), self.func1.indim)
        
        hess = self.cost.Hessian(beta)
        self.assertEqual(np.shape(hess) [0], self.func0.indim)
        self.assertEqual(np.shape(hess) [0], self.func1.indim)
        self.assertEqual(np.shape(hess) [1], self.func0.indim)
        self.assertEqual(np.shape(hess) [1], self.func1.indim)

    def test_dense(self):
        """! Test the dense cost function (clf::DenseCostFunction) python interface"""

        ## The first penalty function
        self.func0 = DensePenaltyFunctionTest0()
        ## The second penalty function
        self.func1 = DensePenaltyFunctionTest1()
        ## The cost function
        self.cost = clf.DenseCostFunction([self.func0, self.func1])

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(self.cost.InputDimension())])
        jac = self.cost.Jacobian(beta)
        self.assertEqual(np.shape(jac) [1], self.func0.indim)
        self.assertEqual(np.shape(jac) [1], self.func1.indim)
        self.assertEqual(np.shape(jac) [0], self.func0.outdim+self.func1.outdim)
        self.assertAlmostEqual(np.linalg.norm(jac[0:self.func0.outdim, 0:self.func0.indim] - self.func0.Jacobian(beta)), 0.0)
        self.assertAlmostEqual(np.linalg.norm(jac[self.func0.outdim:self.func0.outdim+self.func1.outdim, 0:self.func1.indim] - self.func1.Jacobian(beta)), 0.0)

    def test_sparse(self):
        """! Test the sparse cost function (clf::SparseCostFunction) python interface"""

        ## The first penalty function
        self.func0 = SparsePenaltyFunctionTest0()
        ## The second penalty function
        self.func1 = SparsePenaltyFunctionTest1()
        ## The cost function
        self.cost = clf.SparseCostFunction([self.func0, self.func1])

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(self.cost.InputDimension())])
        jac = self.cost.Jacobian(beta)
        self.assertEqual(np.shape(jac) [1], self.func0.indim)
        self.assertEqual(np.shape(jac) [1], self.func1.indim)
        self.assertEqual(np.shape(jac) [0], self.func0.outdim+self.func1.outdim)
        self.assertAlmostEqual(scipy.sparse.linalg.norm(jac[0:self.func0.outdim, 0:self.func0.indim] - self.func0.Jacobian(beta)), 0.0)
        self.assertAlmostEqual(scipy.sparse.linalg.norm(jac[self.func0.outdim:self.func0.outdim+self.func1.outdim, 0:self.func1.indim] - self.func1.Jacobian(beta)), 0.0)
